File: prep/implementation/iterations/iteration_cpp/trie.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

struct TrieNode {
  int score;
  map<char, TrieNode *> children;

  TrieNode() { this->score = 0; }

  ~TrieNode() {
    for (auto pair : children) {
      delete pair.second;
    }
  }
};

class Trie {
private:
  TrieNode *root;

  void insert(TrieNode *node, string word, int score) {
    for (char c : word) {
      if (node->children.find(c) == node->children.end()) {
        node->children[c] = new TrieNode();
      }
      node = node->children[c];
      node->score = max(node->score, score);
    }
  }

  int search(TrieNode *node, string word) {
    for (char c : word) {
      if (node->children.find(c) == node->children.end()) {
        return -1;
      }
      node = node->children[c];
    }

    return node->score;
  }

public:
  Trie() { this->root = new TrieNode(); }
  ~Trie() { delete this->root; }

  void insert(string word, int score) { insert(root, word, score); }

  int search(string word) { return search(root, word); }
};

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, nq;
  cin >> n >> nq;

  Trie trie;
  string word;
  int score;

  for (int i = 0; i < n; i++) {
    cin >> word >> score;
    trie.insert(word, score);
  }

  while (nq--) {
    cin >> word;
    cout << trie.search(word) << "\n";
  }

  return 0;
}

File: prep/implementation/main_iterations/iteration4/Trie.java
import java.util.*;

public class Trie {
  static class TrieNode {
    boolean isTerminal;
    int score;
    Map<Character, TrieNode> children;

    TrieNode() {
      this.children = new HashMap<>();
      this.score = -1;
    }
  }

  private TrieNode root;

  public Trie() { this.root = new TrieNode(); }

  public void insert(String s, int score) { insert(this.root, s, score); }

  private void insert(TrieNode node, String s, int score) {
    for (char c : s.toCharArray()) {
      if (node.children.get(c) == null) {
        node.children.put(c, new TrieNode());
      }
      node = node.children.get(c);
      node.score = Math.max(node.score, score);
    }
    node.isTerminal = true;
  }

  public int search(String s) { return search(this.root, s); }

  private int search(TrieNode node, String s) {
    for (char c : s.toCharArray()) {
      if (node.children.get(c) == null) {
        return -1;
      }
      node = node.children.get(c);
    }

    return node.score;
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int nq = in.nextInt();
      in.nextLine();

      Trie trie = new Trie();
      for (int i = 0; i < n; i++) {
        String[] inp = in.nextLine().trim().split(" ");
        trie.insert(inp[0], Integer.parseInt(inp[1]));
      }

      while (nq-- > 0) {
        String word = in.nextLine().trim();
        System.out.println(trie.search(word));
      }
    }
  }
}